atc/db/access_token.go

```diff
@@ -0,0 +1,60 @@
+package db
+
+import (
+	"database/sql/driver"
+	"encoding/json"
+	"fmt"
+
+	"gopkg.in/square/go-jose.v2/jwt"
+)
+
+type AccessToken struct {
+	Token  string
+	Claims Claims
+}
+
+func scanAccessToken(rcv *AccessToken, scan scannable) error {
+	return scan.Scan(&rcv.Token, &rcv.Claims)
+}
+
+type Claims struct {
+	jwt.Claims
+	FederatedClaims `json:"federated_claims"`
+	RawClaims       map[string]interface{} `json:"-"`
+}
+
+type FederatedClaims struct {
+	Username  string `json:"user_name"`
+	Connector string `json:"connector_id"`
+}
+
+func (c Claims) MarshalJSON() ([]byte, error) {
+	return json.Marshal(c.RawClaims)
+}
+
+func (c *Claims) UnmarshalJSON(data []byte) error {
+	type target Claims
+	var t target
+	if err := json.Unmarshal(data, &t); err != nil {
```

Comment:
This was neat to dig into. It wasn't obvious to me why Claims had to be redefined until I removed`target` and got a stack trace showing the recursive call to `Claims.UnmarshalJSON` that then occurs!

Reply:
Yup! Learned that trick from @vito's steps refactor:

https://github.com/concourse/concourse/blob/a822574e7d1e3e30a54d8cac416bc829c0103d02/atc/steps.go#L414-L415